File: src/routes/posts.py
from flask import Blueprint, request, jsonify
from sqlalchemy import desc
from src.models.user import db, Post, PostLike, Comment
from src.routes.auth import require_auth

posts_bp = Blueprint('posts', __name__)

@posts_bp.route('/posts', methods=['GET'])
@require_auth
def get_posts():
    try:
        # 獲取所有帖子，按點讚數降序，然後按創建時間降序排序
        posts = Post.query.order_by(desc(Post.likes_count), desc(Post.created_at)).all()
        
        current_user_id = request.current_user.id
        posts_data = [post.to_dict(current_user_id) for post in posts]
        
        return jsonify({'posts': posts_data}), 200

    except Exception as e:
        return jsonify({'error': f'獲取帖子失敗: {str(e)}'}), 500

@posts_bp.route('/posts', methods=['POST'])
@require_auth
def create_post():
    try:
        data = request.get_json()
        title = data.get('title', '').strip()
        content = data.get('content', '').strip()

        if not title or not content:
            return jsonify({'error': '請填寫帖子標題和內容'}), 400

        if len(title) > 200:
            return jsonify({'error': '標題不能超過200個字符'}), 400

        if len(content) > 10000:
            return jsonify({'error': '內容不能超過10000個字符'}), 400

        # 創建新帖子
        post = Post(
            title=title,
            content=content,
            user_id=request.current_user.id
        )

        db.session.add(post)
        db.session.commit()

        return jsonify({
            'message': '帖子發布成功',
            'post': post.to_dict(request.current_user.id)
        }), 201

    except Exception as e:
        db.session.rollback()
        return jsonify({'error': f'發布帖子失敗: {str(e)}'}), 500

@posts_bp.route('/posts/<int:post_id>', methods=['GET'])
@require_auth
def get_post(post_id):
    try:
        post = Post.query.get_or_404(post_id)
        current_user_id = request.current_user.id
        
        return jsonify({'post': post.to_dict(current_user_id)}), 200

    except Exception as e:
        return jsonify({'error': f'獲取帖子失敗: {str(e)}'}), 500

@posts_bp.route('/posts/<int:post_id>/like', methods=['POST'])
@require_auth
def toggle_post_like(post_id):
    try:
        post = Post.query.get_or_404(post_id)
        user_id = request.current_user.id

        # 檢查是否已經點讚
        existing_like = PostLike.query.filter_by(post_id=post_id, user_id=user_id).first()

        if existing_like:
            # 取消點讚
            db.session.delete(existing_like)
            post.likes_count = max(0, post.likes_count - 1)
            action = 'unliked'
        else:
            # 添加點讚
            like = PostLike(post_id=post_id, user_id=user_id)
            db.session.add(like)
            post.likes_count += 1
            action = 'liked'

        db.session.commit()

        return jsonify({
            'message': f'帖子{action}',
            'likes_count': post.likes_count,
            'liked': action == 'liked'
        }), 200

    except Exception as e:
        db.session.rollback()
        return jsonify({'error': f'點讚操作失敗: {str(e)}'}), 500

@posts_bp.route('/posts/<int:post_id>/comments', methods=['GET'])
@require_auth
def get_post_comments(post_id):
    try:
        post = Post.query.get_or_404(post_id)
        
        # 獲取評論，按點讚數降序，然後按創建時間降序排序
        comments = Comment.query.filter_by(post_id=post_id).order_by(
            desc(Comment.likes_count), desc(Comment.created_at)
        ).all()
        
        current_user_id = request.current_user.id
        comments_data = [comment.to_dict(current_user_id) for comment in comments]
        
        return jsonify({'comments': comments_data}), 200

    except Exception as e:
        return jsonify({'error': f'獲取評論失敗: {str(e)}'}), 500

@posts_bp.route('/posts/<int:post_id>/comments', methods=['POST'])
@require_auth
def create_comment(post_id):
    try:
        post = Post.query.get_or_404(post_id)
        
        data = request.get_json()
        content = data.get('content', '').strip()

        if not content:
            return jsonify({'error': '請填寫評論內容'}), 400

        if len(content) > 5000:
            return jsonify({'error': '評論內容不能超過5000個字符'}), 400

        # 創建新評論
        comment = Comment(
            content=content,
            post_id=post_id,
            user_id=request.current_user.id
        )

        # 更新帖子的評論數
        post.comments_count += 1

        db.session.add(comment)
        db.session.commit()

        return jsonify({
            'message': '評論發表成功',
            'comment': comment.to_dict(request.current_user.id)
        }), 201

    except Exception as e:
        db.session.rollback()
        return jsonify({'error': f'發表評論失敗: {str(e)}'}), 500

File: src/routes/posts.ts
import { Hono } from 'hono';
import { HTTPException } from 'hono/http-exception';
import { Post } from '../models';
import { requireAdmin } from './auth';

export const postRoutes = new Hono<{ Bindings: Env }>();

// Get all posts with pagination and sorting by likes
postRoutes.get('/', async (c) => {
  const page = parseInt(c.req.query('page') || '1');
  const limit = parseInt(c.req.query('limit') || '10');
  const offset = (page - 1) * limit;

  const db = c.env.DB;
  const posts = await db.prepare(`
    SELECT p.*, u.username 
    FROM posts p 
    JOIN users u ON p.user_id = u.id 
    ORDER BY p.likes_count DESC, p.created_at DESC 
    LIMIT ? OFFSET ?
  `).bind(limit, offset).all();

  return c.json({ posts: posts.results });
});

// Get a specific post
postRoutes.get('/:id', async (c) => {
  const postId = parseInt(c.req.param('id'));
  const db = c.env.DB;

  const post = await db.prepare(`
    SELECT p.*, u.username 
    FROM posts p 
    JOIN users u ON p.user_id = u.id 
    WHERE p.id = ?
  `).bind(postId).first<Post>();

  if (!post) {
    throw new HTTPException(404, { message: 'Post not found' });
  }

  return c.json({ post });
});

// Create a new post
postRoutes.post('/', async (c) => {
  const { title, content } = await c.req.json();
  const user = c.get('user');

  if (!title || !content) {
    throw new HTTPException(400, { message: 'Title and content are required' });
  }

  const db = c.env.DB;
  const { success, results } = await db.prepare(
    'INSERT INTO posts (title, content, user_id) VALUES (?, ?, ?)'
  ).bind(title, content, user.id).run();

  if (!success) {
    throw new HTTPException(500, { message: 'Failed to create post' });
  }

  return c.json({ message: 'Post created successfully', postId: results.lastInsertRowId }, 201);
});

// Update a post (only by the author)
postRoutes.put('/:id', async (c) => {
  const postId = parseInt(c.req.param('id'));
  const { title, content } = await c.req.json();
  const user = c.get('user');

  if (!title || !content) {
    throw new HTTPException(400, { message: 'Title and content are required' });
  }

  const db = c.env.DB;
  const post = await db.prepare('SELECT * FROM posts WHERE id = ?').bind(postId).first();

  if (!post) {
    throw new HTTPException(404, { message: 'Post not found' });
  }

  if (post.user_id !== user.id) {
    throw new HTTPException(403, { message: 'You can only edit your own posts' });
  }

  const { success } = await db.prepare(
    'UPDATE posts SET title = ?, content = ?, last_modified_at = CURRENT_TIMESTAMP WHERE id = ?'
  ).bind(title, content, postId).run();

  if (!success) {
    throw new HTTPException(500, { message: 'Failed to update post' });
  }

  const updatedPost = await db.prepare(`
    SELECT p.*, u.username 
    FROM posts p 
    JOIN users u ON p.user_id = u.id 
    WHERE p.id = ?
  `).bind(postId).first<Post>();

  return c.json({ message: 'Post updated successfully', post: updatedPost });
});

// Delete a post (admin only)
postRoutes.delete('/:id', requireAdmin, async (c) => {
  const postId = parseInt(c.req.param('id'));
  const db = c.env.DB;

  const post = await db.prepare('SELECT * FROM posts WHERE id = ?').bind(postId).first();

  if (!post) {
    throw new HTTPException(404, { message: 'Post not found' });
  }

  // Delete related comments and likes first
  await db.prepare('DELETE FROM comment_likes WHERE comment_id IN (SELECT id FROM comments WHERE post_id = ?)').bind(postId).run();
  await db.prepare('DELETE FROM comments WHERE post_id = ?').bind(postId).run();
  await db.prepare('DELETE FROM post_likes WHERE post_id = ?').bind(postId).run();
  
  const { success } = await db.prepare('DELETE FROM posts WHERE id = ?').bind(postId).run();

  if (!success) {
    throw new HTTPException(500, { message: 'Failed to delete post' });
  }

  return c.json({ message: 'Post deleted successfully' });
});

// Like/unlike a post
postRoutes.post('/:id/like', async (c) => {
  const postId = parseInt(c.req.param('id'));
  const user = c.get('user');
  const db = c.env.DB;

  const post = await db.prepare('SELECT * FROM posts WHERE id = ?').bind(postId).first();
  if (!post) {
    throw new HTTPException(404, { message: 'Post not found' });
  }

  const existingLike = await db.prepare('SELECT * FROM post_likes WHERE post_id = ? AND user_id = ?')
    .bind(postId, user.id).first();

  if (existingLike) {
    // Unlike
    await db.prepare('DELETE FROM post_likes WHERE post_id = ? AND user_id = ?').bind(postId, user.id).run();
    await db.prepare('UPDATE posts SET likes_count = likes_count - 1 WHERE id = ?').bind(postId).run();
    return c.json({ message: 'Post unliked', liked: false });
  } else {
    // Like
    await db.prepare('INSERT INTO post_likes (post_id, user_id) VALUES (?, ?)').bind(postId, user.id).run();
    await db.prepare('UPDATE posts SET likes_count = likes_count + 1 WHERE id = ?').bind(postId).run();
    return c.json({ message: 'Post liked', liked: true });
  }
});

// Get comments for a post
postRoutes.get('/:id/comments', async (c) => {
  const postId = parseInt(c.req.param('id'));
  const db = c.env.DB;

  const comments = await db.prepare(`
    SELECT c.*, u.username 
    FROM comments c 
    JOIN users u ON c.user_id = u.id 
    WHERE c.post_id = ? 
    ORDER BY c.likes_count DESC, c.created_at ASC
  `).bind(postId).all();

  return c.json({ comments: comments.results });
});

// Add a comment to a post
postRoutes.post('/:id/comments', async (c) => {
  const postId = parseInt(c.req.param('id'));
  const { content } = await c.req.json();
  const user = c.get('user');

  if (!content) {
    throw new HTTPException(400, { message: 'Content is required' });
  }

  const db = c.env.DB;
  const post = await db.prepare('SELECT * FROM posts WHERE id = ?').bind(postId).first();
  if (!post) {
    throw new HTTPException(404, { message: 'Post not found' });
  }

  const { success } = await db.prepare(
    'INSERT INTO comments (content, post_id, user_id) VALUES (?, ?, ?)'
  ).bind(content, postId, user.id).run();

  if (!success) {
    throw new HTTPException(500, { message: 'Failed to create comment' });
  }

  // Update post comment count
  await db.prepare('UPDATE posts SET comments_count = comments_count + 1 WHERE id = ?').bind(postId).run();

  return c.json({ message: 'Comment created successfully' }, 201);
});



